packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -660,43 +666,60 @@ class TextSelectionOverlay {
       offset: lineAtOffset.baseOffset,
     );
 
-    final Rect lineBoundaries = Rect.fromPoints(
+    final Rect localLineBoundaries = Rect.fromPoints(
       renderEditable.getLocalRectForCaret(positionAtBeginningOfLine).topCenter,
       renderEditable.getLocalRectForCaret(positionAtEndOfLine).bottomCenter,
     );
+    final Rect globalLineBoundaries = Rect.fromPoints(
+      renderEditable.localToGlobal(localLineBoundaries.topLeft),
+      renderEditable.localToGlobal(localLineBoundaries.bottomRight),
+    );
+
+    final Rect globalCaretRect = Rect.fromPoints(
+        renderEditable.localToGlobal(localCaretRect.topLeft),
+        renderEditable.localToGlobal(localCaretRect.bottomRight),
+    );
 
     return MagnifierInfo(
-      fieldBounds: globalRenderEditableTopLeft & renderEditable.size,
-      globalGesturePosition: globalGesturePosition,
-      caretRect: localCaretRect.shift(globalRenderEditableTopLeft),
-      currentLineBoundaries: lineBoundaries.shift(globalRenderEditableTopLeft),
+      fieldBounds: Rect.fromPoints(globalRenderEditableTopLeft, globalRenderEditableBottomRight),
+      globalGesturePosition: overlayGesturePosition,
+      caretRect: globalCaretRect,
+      currentLineBoundaries: globalLineBoundaries,
     );
   }
 
-  // The contact position of the gesture at the current end handle location.
-  // Updated when the handle moves.
+  // The contact position of the gesture at the current end handle location, in
+  // global coordinates. Updated when the handle moves.
   late double _endHandleDragPosition;
 
   // The distance from _endHandleDragPosition to the center of the line that it
-  // corresponds to.
+  // corresponds to, in global coordinates.
   late double _endHandleDragPositionToCenterOfLine;
 
   void _handleSelectionEndHandleDragStart(DragStartDetails details) {
     if (!renderObject.attached) {
       return;
     }
 
-    // This adjusts for the fact that the selection handles may not
-    // perfectly cover the TextPosition that they correspond to.
     _endHandleDragPosition = details.globalPosition.dy;
-    final Offset endPoint =
-        renderObject.localToGlobal(_selectionOverlay.selectionEndpoints.last.point);
-    final double centerOfLine = endPoint.dy - renderObject.preferredLineHeight / 2;
-    _endHandleDragPositionToCenterOfLine = centerOfLine - _endHandleDragPosition;
+
+    // Use local coordinates when dealing with line height. because in case of a
+    // scale transformation, the line height will also be scaled.
+    final double centerOfLineLocal = _selectionOverlay.selectionEndpoints.last.point.dy
+        - renderObject.preferredLineHeight / 2;
+    final double centerOfLineGlobal = renderObject.localToGlobal(
+      Offset(0.0, centerOfLineLocal),
+    ).dy;
+    _endHandleDragPositionToCenterOfLine = centerOfLineGlobal  - details.globalPosition.dy;
+    // This adjusts for the fact that the selection handles may not perfectly
```

Comment:
`This` refers to using `seslecionEndpoints` instead of the pointer event? Then maybe move the comment up?

Reply:
I rephrased the comment to be more clear about what it's referring to.